cluster-autoscaler/expander/priority/priority.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priority
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+
+	"gopkg.in/yaml.v2"
+
+	"k8s.io/autoscaler/cluster-autoscaler/expander"
+	"k8s.io/autoscaler/cluster-autoscaler/expander/random"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+type priority struct {
+	fallbackStrategy expander.Strategy
+	changesChan      <-chan watch.Event
+	priorities       map[int][]*regexp.Regexp
+	padlock          sync.RWMutex
+	okConfigUpdates  int
+	badConfigUpdates int
+}
+
+// NewStrategy returns an expansion strategy that picks node groups based on user-defined priorities
+func NewStrategy(initialPriorities string, priorityChangesChan <-chan watch.Event) (expander.Strategy, errors.AutoscalerError) {
+	res := &priority{
+		fallbackStrategy: random.NewStrategy(),
+		changesChan:      priorityChangesChan,
+	}
+	if err := res.parsePrioritiesYAMLString(initialPriorities); err != nil {
+		return nil, errors.ToAutoscalerError(errors.InternalError, err)
+	}
+	go func() {
+		// TODO: how to terminate on process shutdown?
+		for event := range priorityChangesChan {
+			cm, ok := event.Object.(*apiv1.ConfigMap)
+			if !ok {
+				klog.Exit("Unexpected object type received on the configmap update channel in priority expander")
+			}
+
+			if event.Type == watch.Deleted {
+				klog.Warning("Configmap for priority expander was deleted, no updates will be processed until recreated.")
+				res.badConfigUpdates++
+				continue
+			}
+
+			prioString, found := cm.Data[ConfigMapKey]
+			if !found {
+				klog.Warningf("Wrong configmap for priority expander, doesn't contain %s key. Ignoring update.", ConfigMapKey)
+				res.badConfigUpdates++
+				continue
+			}
+			if err := res.parsePrioritiesYAMLString(prioString); err != nil {
+				klog.Warningf("Wrong configuration for priority expander: %v. Ignoring update.", err)
```

Comment:
I would also send an event, update status config map, etc. Be extremely noisy in case of misconfiguriation.